### ISSUE TITLE
Idea: Runtimes appended to AppImages

### ISSUE THREAD

**Issue by probonopd** [AppImage/AppImageKit#890]
Just some food for thought for future consideration. Not something to be rushed.
I shared this idea on IRC today:

```
it seems that some people like the idea of having runtimes, so that the AppImage + the runtime = no dependencies on the system other than the kernel
so far, i was heavily opposed to it. but here is some food for thought:
what if we _appended_ the required runtime to each AppImage
another squashfs
which gets downloaded if you regularly download
which can also be left away (e.g., HTTP Range Request) if you already have/dont't want it
the default is that you get it, but you can opt out
something like that
just an idea for now
(might become AppImage image type 3 one day?)
Example:
https://github.com/Hackerl/Wine_Appimage/releases
the AppImages there all need Wine-x86_64.AppImage to run
they need nothing else from the system
they have their own glibc and ld-linux
i imagine that we could have one KDG, GNOME, KDE, WINE,... "RuntimeImage" per year or so
that people could depend on by appending it
who downloads it with a browser downloads the whole thing, who downloads it with something like AppImageUpdate would only get what they need
but as i said, just a vague idea at this point
(just had the idea after having argued with distro people how bad AppImage is because it lacks runtimes yesterday, and having seen what our Chinese friends do with AppImage and WINE today)
```

**Comment by probonopd** [comment]
Yes, the runtime could e.g., be specified based on its SHA-1 hash or something along those lines, to make _really_ sure it matches.

I think there are still a couple of issues with this idea:
* What happens if user A has AppImage One with the appended runtime Foo
* The same user A then downloads AppImage Two that also requires runtime Foo
* At some time, it is desirable to have deduplication (e.g., a "smart" download tool would not download runtime Foo a second time), and not save runtime Foo twice on the filesystem
* If the user now copies AppImage Two (which does not contain runtime Foo because it is already included in AppImage One) to a friend, then the friend is lost because he has an AppImage that does not run without runtime Foo (which he does not have)

Maybe this can be solved in a clever way, but it is not obvious how.

**Comment by shoogle** [comment]
> Maybe this can be solved in a clever way, but it is not obvious how.

I don't think this is a case where you can't have it both ways, so you need to look at the most likely usage patterns.

- If the first person downloaded AppImage Two with a browser then there's no issue.
- If they downloaded it using a special manager program then they are unlikely be using AppImages as portable files. They wouldn't expect it to work outside of their own Applications folder.

If you are really desperate to solve this, then the only way I can think of is for AppImages to say "I need this program from this URL to be able to run" and the manager program (if installed) would download it for them when they attemp to run an AppImage for the first time.

Ultimately, however, people don't pass software around on USB drives as it's much easier just to send your friend a URL where they can download a copy for themselves.

**Comment by probonopd** [comment]
> people don't pass software around on USB drives

In fact, some do. Air-gapped machines. It's a core idea of AppImage that you don't need to be online at the moment in which you receive an AppImage. I was more contemplating hardlinks, block-based deduplication and such things, but have not found a clever solution yet.

**Comment by shoogle** [comment]
The more interesting question, in my opinion, is what to do when AppImage One is downloaded. Should runtime Foo remain appended to it, or should it be moved into a neutral location like `~/Libraries` or `~/Applications/Libraries`?

My vote would be to move it, as this has the following advantages:

- The runtime remains available for AppImage Two even if AppImage One is deleted.
- If it wasn't moved app developers might think that the runtime will always be appended and write their software based on this assumption. Moving the runtime forces them to consider this scenario.

**Comment by probonopd** [comment]
We could also let it appended until the user does some explicit action to remove it. Like the [`lipo`](https://www.unix.com/man-page/osx/1/lipo/) command on the Mac to trim unneeded architectures from "fat binary" binaries that contain multiple architectures.

Once the user does that (or a partial download using Range Requests), we could mark the AppImage as "slimmed".

**Comment by probonopd** [comment]
Another consideration: AppImages typically are _much_ smaller than the download sizes of runtime-based systems when the MB for the application and its runtime(s) are added.

* The full Qt/KDE library set is XXX MB whereas the subset needed for a particular application may be just 10-20 MB
* The full WINE library set is XXX MB whereas the subset needed for a particular application may be just a few MB

So not really sure it's worth it for those.

Even Java has some effort undergoing to be be able to bundle just the subset that is actually needed by a particular application.

**Comment by shoogle** [comment]
I wouldn't recommend stripping out parts of the libraries, other than those that are already recognised by [`linuxdeploy-plugin-qt`](https://github.com/linuxdeploy/linuxdeploy-plugin-qt/). A better idea would be to statically link the libraries so that unnecessary symbols are stipped by the compiler.

I've looked at doing this in the past but it has always seemed to be more trouble than it's worth as all the distributions are heavily invested in dynamic linking. However, if there was an easy way to fetch libraries and compile them statically (maybe via something like [Linuxbrew](https://linuxbrew.sh/)) then this could be worth considering.

Even then, however, it might be best to statically link small dependencies and dynamically link the larger runtimes.

**Comment by probonopd** [comment]
With "slimming" I didn't mean stripping or removing symbols from libraries, but in our case removing the secondary, appended squashfs image from the AppImage.

**Comment by shoogle** [comment]
Yes, I understood that. I was referring to your [next comment](https://github.com/AppImage/AppImageKit/issues/890#issuecomment-445958192) which talks about only shipping a *subset* of a library/runtime, presumably by deleting the unused `.so` files from the runtime before building the AppImage (i.e. manual stripping). I'm saying it would be better to let the compiler do the stripping automatically via static linking. Both of these are methods are ways to make the AppImage small enough that "slimming" (as in separating the app and the runtime) is unnecessary.

**Comment by probonopd** [comment]
Static linking may not be allowable without shipping object files for license reasons, e.g., for Qt.

**Comment by TheAssassin** [comment]
Horrible idea to allow AppImages to work _without_ said "runtime" image. Providing such base images in general might be a good idea, allowing people to get started or update base libraries more often, but they must be shipped together in the same file as the original AppImage. Otherwise, this breaks "one app = one file", and there is no way to avoid this. And this is the key feature of AppImage.

**Comment by probonopd** [comment]
Yes, this is why I am really hesitant to do anything in this direction.
